staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -512,6 +512,62 @@ func PollImmediateUntil(interval time.Duration, condition ConditionFunc, stopCh
 	}
 }
 
+// PollUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollUntilWithContext always waits interval before the first run of 'condition'.
+// 'condition' will always be invoked at least once.
+func PollUntilWithContext(interval time.Duration, condition ConditionFunc, ctx context.Context) error {
```

Comment:
@wojtek-t I am copy pasting your comment here, so we can continue here -

"Actually - I think I'm against introducing this function.

Why can't just the caller do

PollUntil(interval, condition, ctx.Done())
?
"

Reply:
`PollUntil` will derive a context from `stopCh` using `contextForChannel`. it seems wasteful to derive a new context and also to spin up a new goroutine to track the newly derived context. 

```
func contextForChannel(parentCh <-chan struct{}) (context.Context, context.CancelFunc) {
	ctx, cancel := context.WithCancel(context.Background())

	go func() {
		select {
		case <-parentCh:
			cancel()
		case <-ctx.Done():
		}
	}()
	return ctx, cancel
}
```

Reply:
@wojtek-t  I agree with your comment on the auxiliary functions, it's a maintenance burden. I think we need to find a way to retire or deprecate these functions and pave the way for `context` bound functions (our clients already take `context` and it seems a natural extension to use `context`) 
Also, I am hoping in the future we will convert `PostStartHook` `StopCh` to a `context`.  But this is outside the scope of this PR :)


I used `context` thinking that, new functions should be  `context` bound instead of a channel. 

I think returning a `channel` with `<- chan struct{}` takes away the ability for the caller to explicitly cancel, and returning `chan struct{}` is not idiomatic.

Let me know your thoughts, if you have a strong opinion on this I am happy to return a channel.